speechbrain/nnet/ligru/cuda_ligru_cell.py

```diff
@@ -0,0 +1,223 @@
+"""This library implements a LiGRU cell needed for the LiGRU.
+
+LiGRU is a Gated Recurrent Unit introduced by Assoc. Prof Mirco 
+Ravanelli in 2018 (see: https://arxiv.org/pdf/1803.10225.pdf).
+
+Authors
+ * Adel Moumen 2022
+"""
+
+import torch
+import torch.autograd as autograd
+
+try:
+    import cupy as cp
+except ImportError:
+    err_msg = "The optional dependency CuPy is needed to use LiGRU on CuPy\n"
+    err_msg += "Cannot import CuPy.\n"
+    err_msg += "Please follow the instructions below\n"
+    err_msg += "=============================\n"
+    err_msg += "If you use your localhost:\n"
+    err_msg += "$ python -m pip install -U setuptools pip\n"
+    err_msg += "$ pip install cupy-cudaXXX (XXX is your Cuda Toolkit version)\n"
+    err_msg += "If you use conda:\n"
+    err_msg += "$ conda install -c conda-forge cupy"
+    err_msg += "for more details: https://docs.cupy.dev/en/stable/install.html"
+    err_msg += "=============================\n"
+    raise ImportError(err_msg)
+
+_preamble_gradient_relu = '''
+template <typename T> __device__ T gradient_activation_hcand(T x) { return (x > 0. ? 1. : 0.); }
+'''
+
+_preamble_gradient_leaky_relu = '''
+template <typename T> __device__ T gradient_activation_hcand(T x) { 
+    T negative_slope = 1e-2;  
+    return (x > 0. ? 1. : negative_slope); 
+}
+'''
+
+_preamble_gradient_tanh = '''
+template <typename T> __device__ T gradient_activation_hcand(T x) { return 1 - x * x; }
+'''
+
+def transform_tensor_to_cupy(x):
+    """Transform a PyTorch Tensor located on device="cuda" to a CuPy array. 
+    
+    Argument
+    --------
+        x : torch.Tensor
+    """
+    return cp.ascontiguousarray(cp.from_dlpack(torch.utils.dlpack.to_dlpack(x.detach())))
+
+
+
+class _ligru_cell_jit(torch.nn.Module):
+    """This class redefines the forward of a LiGRU cell.
+    """
+    def __init__(self, act):
+        super(_ligru_cell_jit, self).__init__()
+        self.act = act 
+
+    def forward(self, wx, u, ht, drop_mask):
+        """Compute the forward pass of a LiGRU cell.         
+        
+        Arguments
+        ---------
+        wx : torch.Tensor
+            Linearly transformed input. 
+        u  : torch.Tensor
+            Recurrent weight. 
+        ht : torch.Tensor
+            Hidden state. 
+        drop_mask : torch.Tensor
+            Dropout mask. 
+        act : nn.Module
+            Activation Function. 
+        """
+        # save values for backward
+        hiddens        = []
```

Comment:
I also prefer this style of alignement with Python. HOWEVER, this is not compliant with what SpeechBrain has adopted :-P "=" symbols must have one space before and after